### PR TITLE
API Un-deploy event notify to control plane

### PR DESCRIPTION
**Pull request by shanakama**
### Purpose
$Subject

### Issues

Fixes # https://github.com/wso2/product-microgateway/issues/2528

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/notifier/deployment_notifier.go [hunk 1]
```diff
@@ -82,3 +83,51 @@ func SendRevisionUpdate(deployedRevisionList []*DeployedAPIRevision) {
...
+		retries++
+		req, _ := http.NewRequest("POST", revisionEP, bytes.NewBuffer(jsonValue))
+		req.Header.Set("Authorization", basicAuth)
+		req.Header.Set("Content-Type", "application/json")
+		resp, err := tlsutils.InvokeControlPlane(req, cpConfigs.SkipSSLVerification)
+
+		success := true
+		if err != nil {
+			logger.LoggerNotifier.Warnf("Error response from %v for attempt %v : %v", revisionEP, retries, err.Error())
+			success = false
```
Comment (VirajSalaka) [inline adapter/internal/notifier/deployment_notifier.go:122]: needs continue here ?

**Comment by renuka-fernando** [comment]
@shanakama shall we do the same changes for the main branch
And if this is not an urgent fix, let's do changes only to the main branch, PRs to the choreo branch is not recommended.

File: adapter/internal/discovery/xds/server.go [hunk 2]
```diff
@@ -598,6 +598,9 @@ func DeleteAPIWithAPIMEvent(uuid, name, version string, environments []string, o
 			for _, environment := range environments {
 				// delete environment if exists
 				delete(apiUUIDToGatewayToVhosts[uuid], environment)
+				if revisionUUID != "" {
+					notifier.SendRevisionUndeploy(uuid, revisionUUID, environment)
```
Comment (pubudu538) [inline adapter/internal/discovery/xds/server.go:602]: Let's call this only if Global Adapter is enabled.

Comment (shanakama) [reply adapter/internal/discovery/xds/server.go:616]: Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/cf1de67208492e8ac71fe95e32a52cf036ab3ee7

File: adapter/internal/notifier/deployment_notifier.go [hunk 3]
```diff
@@ -82,3 +83,51 @@ func SendRevisionUpdate(deployedRevisionList []*DeployedAPIRevision) {
...
+		Environment:  environment,
+	}
+
+	jsonValue, _ := json.Marshal(removedRevision)
+	basicAuth := "Basic " + auth.GetBasicAuth(cpConfigs.Username, cpConfigs.Password)
+	retries := 0
+	for retries < 3 {
+		retries++
+		req, _ := http.NewRequest("POST", revisionEP, bytes.NewBuffer(jsonValue))
+		req.Header.Set("Authorization", basicAuth)
```
Comment (VirajSalaka) [inline adapter/internal/notifier/deployment_notifier.go:115]: isn't there any constants used for these (while introducing deployRevision Update)?

File: adapter/internal/notifier/deployment_notifier.go [hunk 4]
```diff
@@ -82,3 +83,51 @@ func SendRevisionUpdate(deployedRevisionList []*DeployedAPIRevision) {
...
+
+		success := true
+		if err != nil {
+			logger.LoggerNotifier.Warnf("Error response from %v for attempt %v : %v", revisionEP, retries, err.Error())
+			success = false
+		}
+		if resp != nil && resp.StatusCode != http.StatusOK {
+			logger.LoggerNotifier.Warnf("Error response status code %v from %v for attempt %v", resp.StatusCode, revisionEP, retries)
+			success = false
+		}
```
Comment (VirajSalaka) [inline adapter/internal/notifier/deployment_notifier.go:127]: Would not it be better to have at least one second delay before retrying. Or may be 5 secs.
@AmaliMatharaarachchi  wdyt?

File: adapter/internal/notifier/deployment_notifier.go (hunk 3 above)
Comment (shanakama) [reply adapter/internal/notifier/deployment_notifier.go:115]: there wasn't . better to have. Will fix.

Comment (shanakama) [reply adapter/internal/notifier/deployment_notifier.go]: Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/b551edd52195244f99b9c10e471d9e431125a7c9

File: adapter/internal/notifier/deployment_notifier.go (hunk 4 above)
Comment (shanakama) [reply adapter/internal/notifier/deployment_notifier.go:131]: Added 2 seconds delay.

**Review by VirajSalaka — APPROVED** [review]

File: adapter/internal/notifier/deployment_notifier.go [hunk 5]
```diff
@@ -82,3 +87,52 @@ func SendRevisionUpdate(deployedRevisionList []*DeployedAPIRevision) {
...
+	cpConfigs := conf.ControlPlane
+
+	revisionEP := cpConfigs.ServiceURL
+	if strings.HasSuffix(revisionEP, "/") {
+		revisionEP += unDeployedRevisionEP
+	} else {
+		revisionEP += "/" + unDeployedRevisionEP
+	}
+
+	if apiUUID == "" || revisionUUID == "" || environment == "" {
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/notifier/deployment_notifier.go:103]: Don't we have to check if the control plane is enabled here?

File: adapter/internal/notifier/deployment_notifier.go [hunk 6]
```diff
@@ -82,3 +87,52 @@ func SendRevisionUpdate(deployedRevisionList []*DeployedAPIRevision) {
...
+	for retries < 3 {
+		retries++
+		req, _ := http.NewRequest("POST", revisionEP, bytes.NewBuffer(jsonValue))
+		req.Header.Set(authHeader, basicAuth)
+		req.Header.Set(contentTypeHeader, "application/json")
+		resp, err := tlsutils.InvokeControlPlane(req, cpConfigs.SkipSSLVerification)
+
+		success := true
+		if err != nil {
+			logger.LoggerNotifier.Warnf("Error response from %v for attempt %v : %v", revisionEP, retries, err.Error())
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/notifier/deployment_notifier.go:125]: Isn't it better to set it as an error log?

File: adapter/internal/discovery/xds/server.go [hunk 7]
```diff
@@ -583,9 +583,15 @@ func DeleteAPIsWithUUID(vhost, uuid string, environments []string, organizationI
...
 
 // DeleteAPIWithAPIMEvent deletes API with the given UUID from the given gw environments
-func DeleteAPIWithAPIMEvent(uuid, organizationID string, environments []string) {
+func DeleteAPIWithAPIMEvent(uuid, organizationID string, environments []string, revisionUUID string) {
 	apiIdentifiers := make(map[string]struct{})
 
+	conf, errReadConfig := config.ReadConfigs()
+	if errReadConfig != nil {
+		logger.LoggerOasparser.Fatal("Error loading configuration. ", errReadConfig)
+	}
```
Comment (praminda) [inline adapter/internal/discovery/xds/server.go:592]: we don't need to check this error here. error is only thrown when initial config reading happen from the config file. here its just returning loaded config object.
```suggestion
	conf, _ := config.ReadConfigs()
```

File: adapter/internal/discovery/xds/server.go [hunk 8]
```diff
@@ -602,10 +608,13 @@ func DeleteAPIWithAPIMEvent(uuid, organizationID string, environments []string)
 		} else {
 			// if no error, update internal vhost maps
 			// error only happens when API not found in deleteAPI func
-			logger.LoggerXds.Infof("Successfully undeployed API %v of Organization %v from environments %v", apiIdentifier, organizationID, environments)
+			logger.LoggerXds.Infof("Successfully undeployed API %v of Organization %v from environments %v revision %v before revisionUUID", apiIdentifier, organizationID, environments, revisionUUID)
```
Comment (praminda) [inline adapter/internal/discovery/xds/server.go:611]: meaning of ` revision %v before revisionUUID` confused me. what exactly are we trying to say. is it `Revision xxxx of API Petstore:1.0.0 in org foo was undeployed successfully from abc, def environments`?

Comment (shanakama) [reply adapter/internal/discovery/xds/server.go:611]: Ah sorry, that was a mistakenly committed. `before revisionUUID` this should be removed.

Comment (shanakama) [reply adapter/internal/discovery/xds/server.go]: Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/4d77d188a4aefb79e5d8582c06613d46ae0a1348

File: adapter/internal/discovery/xds/server.go (hunk 7 above)
Comment (shanakama) [reply adapter/internal/discovery/xds/server.go]: Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/4d77d188a4aefb79e5d8582c06613d46ae0a1348

File: adapter/internal/notifier/deployment_notifier.go (hunk 6 above)
Comment (shanakama) [reply adapter/internal/notifier/deployment_notifier.go]: +1 and 
Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/4d77d188a4aefb79e5d8582c06613d46ae0a1348

File: adapter/internal/notifier/deployment_notifier.go (hunk 5 above)
Comment (shanakama) [reply adapter/internal/notifier/deployment_notifier.go]: Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/4d77d188a4aefb79e5d8582c06613d46ae0a1348

**Review by pubudu538 — APPROVED** [review]

File: adapter/internal/notifier/deployment_notifier.go [hunk 9]
```diff
@@ -82,3 +87,52 @@ func SendRevisionUpdate(deployedRevisionList []*DeployedAPIRevision) {
...
+		req.Header.Set(contentTypeHeader, "application/json")
+		resp, err := tlsutils.InvokeControlPlane(req, cpConfigs.SkipSSLVerification)
+
+		success := true
+		if err != nil {
+			logger.LoggerNotifier.Errorf("Error response from %v for attempt %v : %v", revisionEP, retries, err.Error())
+			success = false
+		}
+		if resp != nil && resp.StatusCode != http.StatusOK {
+			logger.LoggerNotifier.Warnf("Error response status code %v from %v for attempt %v", resp.StatusCode, revisionEP, retries)
```
Comment (AmaliMatharaarachchi) [inline adapter/internal/notifier/deployment_notifier.go:129]:
```suggestion
			logger.LoggerNotifier.Errorf("Error response status code %v from %v for attempt %v", resp.StatusCode, revisionEP, retries)
```

Comment (shanakama) [reply adapter/internal/notifier/deployment_notifier.go]: Fixed via : https://github.com/wso2/product-microgateway/pull/2503/commits/a76d06b2a7d9747f501dca91de37cc15b72cb847

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[failed] dev-deployment : 20211215.32

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20211215.43

**Pull request merged**